python/lsst/meas/extensions/scarlet/multiBandDetectionTask.py

```diff
@@ -0,0 +1,268 @@
+        table = SourceTable.make(self.schema, idFactory)
+        sources = SourceCatalog(table)
+        for footprint in footprints:
+            fp_ymin, fp_xmin = footprint.bbox.origin
+            spans, _ = getSpanSetFromImages(
+                footprint.data, xy0=Point2I((xmin + fp_xmin, ymin + fp_ymin))
+            )
+            _footprint = AfwFootprint(spans)
+            source = sources.addNew()
+            source.setFootprint(_footprint)
+
+        if self.config.doDetectPeaks:
+            # Detect peaks in the detection image
+            # and update the footprints in place.
+            peaks = self._detectPeaks(mCoadd, footprints)
+
+        # Attach peaks from an external catalog to the footprints.
+        self._attachExternalPeaks(peaks, footprints, sources, mCoadd.getBBox())
+
+        return pipeBase.Struct(footprints=footprints, sources=sources)
+
+    def _detectPeaks(
+        self,
+        mCoadd: afwImage.MultibandExposure,
+        footprints: Sequence[scl.detect.Footprint],
+    ):
+        """Detect peaks in the detection image"""
+        # Build the noise weighted detection image.
+        sigma = np.median(np.sqrt(mCoadd.variance.array), axis=(1, 2)) / 2
+        detection = np.sum(mCoadd.image.array / sigma[:, None, None], axis=0)
+        # Build a mask from the footprints to remove noise
+        bbox = utils.bboxToScarletBox(mCoadd.getBBox())
+        footprint_image = scl.detect.footprints_to_image(footprints, bbox).data
+        detection *= footprint_image
+        # Convolve the detection image with a larger circular Gaussian PSF
+        model_psf = scl.utils.integrated_circular_gaussian(
+            sigma=self.config.modelPsfSigma
+        )
+        convolved = scipy_convolve(detection, model_psf, mode="same")
+        # Detect peaks in the convolved image
+        xmin, ymin = mCoadd.getBBox().getMin()
+        wide_footprints = scl.detect.detect_footprints(
+            images=convolved,
+            variance=mCoadd.variance.array,
+            origin=(ymin, xmin),
+            min_separation=self.config.minPeakDistance,
+            min_area=self.config.minFootprintArea,
+            peak_thresh=self.config.minPeakSNR,
+            footprint_thresh=self.config.minFootprintSNR,
+            find_peaks=True,
+            remove_high_freq=False,
+            min_pixel_detect=1,
+        )
+        peaks = utils.scarletFootprintsToPeakCatalog(wide_footprints)
+        log.debug("Total number of peaks detected:", len(peaks))
+        return peaks
+
+    def _attachExternalPeaks(
+        self,
+        peaks: PeakCatalog,
+        footprints: Sequence[scl.detect.Footprint],
+        sources: SourceCatalog,
+        bbox: Box2I,
+    ):
+        """Attach peaks to the footprints in the source catalog"""
+        scarletBox = utils.bboxToScarletBox(bbox)
+        xmin, ymin = bbox.getMin()
+        footprint_image = scl.detect.footprints_to_image(footprints, scarletBox).data
```

Comment:
This is a boolean mask of pixels within footprints?

Reply:
No, it's actually an image with an integer (the index of the footprint in the catalog) for each pixel. So `footprint_image[y, x]` gives the index of the footprint at that location. This is actually an idea that Nate and I came up with one day when trying to figure out the best way to insert peaks into existing footprints without having to search the catalogs.

Reply:
Okay, that sounds like a segmentation map then.